Add a Texture View Renderer sample (#302)
opentok/opentok-android-sdk-samples: Texture-Views-Renderer/README.md

# Texture View Renderer

> In this sample we include a custom video renderer based on `android.view.TextureView` class.

Reviewer: ```suggestion
This sample application uses a custom video renderer based on `android.view.TextureView` class.
```

> When it comes to video rendering, there are different alternatives, in the OpenTok SDK, we support two ways of doing it:

Reviewer: ```suggestion
The OpenTok SDK supports multiple ways of video rendering:
```

1. Using a GLSurfaceView
2. Using a TextureView

> Both classes, although both are very similar they have some differences that make them more appropriate according to the use case.

Reviewer: ```suggestion
Although both GLSurfaceView and TextureView are very similar, they have some differences that make them more appropriate according to the use case.
```

> When using GLSurfaceView, it is more complicated to do some view compositions like applying a mask or doing animations with the views, 

Reviewer: ```suggestion
When using GLSurfaceView, it is more complicated to compose views such as  applying a mask or doing animations with the views,
```

> on the other hand, when using TextureViews, these kind of things are feasible.

Reviewer: ```suggestion
however, when using TextureViews, the composition is much easier.
```

By default, OpenTok SDK uses a `GLSurfaceView` to render the video, but you can enable using a `TextureView` when creating a session:

```java
Session builder = new Session.Builder(this, apiKey, sessionId)
    .sessionOptions(new Session.SessionOptions() {
        @Override
        public boolean useTextureViews() {
            return true;
        }
    }).build();
``` 

> In this sample we show how a `TextureView` renderer can be done and how it can be used when rendering video.

Reviewer: ```suggestion
In this sample, we show how a `TextureView` renderer can be implemented and how it can be used when rendering video.
```

## TextureView based renderer

This renderer is an implementation of the OpenTok's `BaseVideoRenderer` and lives in `TextureViewRenderer` class of the sample.

`TextureViewRenderer` uses a `TextureView` android class in order to do the video rendering.

The schema of the `TextureViewRenderer` class is quite simple:

> - When frames are received in the `BaseVideoRenderer.onFrame` method, it is saved in a local variable protected by a lock.

Reviewer: ```suggestion
- When a frame is received in the `BaseVideoRenderer.onFrame` method, it is saved in a local variable protected by a lock.
```

- `TextureView` render thread, takes the frame set by the previous method, and render it using OpenGL fragment and vertex shaders.

Working with OpenGL in Android can be messy given the different versions of OpenGL ES and different Operating system version support.

> To make it easy we have imported some handy classes from Google's [grafika project](https://github.com/google/grafika). 

Reviewer: ```suggestion
To make this easier, we have imported some handy classes from Google's [grafika project](https://github.com/google/grafika). 
```
